---
title: "Untitled"
output: html_document
date: "2023-07-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots
```{bash}
conda activate qiime2-2023.2
```

Create your tutorial directory, and then, move into this directory:
```{bash}
cd ~/workspace
mkdir 16S_workshop/
```


```{bash}
cd 16S_workshop/
```

#Link the tutorial data:
```{bash}

ln -s ~/CourseData/MIC_data/16S_data/raw_data/ .
ln -s ~/CourseData/MIC_data/16S_data/Blueberry_metadata_reduced.tsv .
```

1. First steps
1.1. Inspect raw data
First, letâ€™s take a look at the directory containing our raw reads as well as our metadata file.
```{bash}
ls raw_data
head Blueberry_metadata_reduced.tsv
```

Question 1: How many samples are there?

Question 2: Into what group(s) are the samples classified?

# 1.3. Import FASTQs as QIIME2 artifact
```{bash}
mkdir reads_qza
```



```{bash}
qiime tools import \
    --type SampleData[PairedEndSequencesWithQuality] \
    --input-path raw_data/ \
    --output-path reads_qza/reads.qza \
    --input-format CasavaOneEightSingleLanePerSampleDirFmt
```

# 1.4. Trim primers with cutadapt
```{bash}
## Check the number of cores
lscpu
```


```{bash}
qiime cutadapt trim-paired \
    --i-demultiplexed-sequences reads_qza/reads.qza \
    --p-cores 4 \
    --p-front-f ACGCGHNRAACCTTACC \
    --p-front-r ACGGGCRGTGWGTRCAA \
    --p-discard-untrimmed \
    --p-no-indels \
    --o-trimmed-sequences reads_qza/reads_trimmed.qza
```

# 1.4. (Optional) Summarize trimmed FASTQs
```{bash}
qiime demux summarize \
    --i-data reads_qza/reads_trimmed.qza \
    --o-visualization reads_qza/reads_trimmed_summary.qzv
```

# 2. Denoising the reads into amplicon sequence variants
# 2.1. Join paired-end reads
```{bash}
qiime vsearch merge-pairs \
    --i-demultiplexed-seqs reads_qza/reads_trimmed.qza \
    --o-merged-sequences reads_qza/reads_trimmed_joined.qza
```

# 2.2. Filter out low-quality reads
```{bash}
qiime quality-filter q-score \
    --i-demux reads_qza/reads_trimmed_joined.qza \
    --o-filter-stats filt_stats.qza \
    --o-filtered-sequences reads_qza/reads_trimmed_joined_filt.qza
# 2.3. Summarize joined and filtered reads    
qiime demux summarize \
    --i-data reads_qza/reads_trimmed_joined_filt.qza \
    --o-visualization reads_qza/reads_trimmed_joined_filt_summary.qzv
```

Question 4: How long are our forward reads? Why are there no reverse reads in our file?

Question 5: What would be a good trim length for our reads?

# 2.4. Running Deblur
```{bash}
qiime deblur denoise-16S \
    --i-demultiplexed-seqs reads_qza/reads_trimmed_joined_filt.qza \
    --p-trim-length 392 \
    --p-sample-stats \
    --p-jobs-to-start 4 \
    --p-min-reads 1 \
    --output-dir deblur_output
```

# 2.5. Summarizing deblur output
```{bash}
qiime feature-table summarize \
    --i-table deblur_output/table.qza \
    --o-visualization deblur_output/deblur_table_summary.qzv
```
Question 6: What is the mean sequencing depth per sample after denoising?

Question 7: Which sample has the least reads?

```{bash}

```

# 3. Assign taxonomy to ASVs
## 3.1. (Optional) Build or acquire taxonomic classifier
https://docs.qiime2.org/2022.11/data-resources/ 
```{bash}
```

## 3.2. Run taxonomic classification
```{bash}
cp -r ~/CourseData/MIC_data/16S_data/taxa .
```


```{bash}
#This is the command that you would usually run:

<!-- qiime feature-classifier classify-sklearn \ -->
<!--     --i-reads deblur_output/representative_sequences.qza \ -->
<!--     --i-classifier silva-138-99-nb-classifier.qza \ -->
<!--     --p-n-jobs 4 \ -->
<!--     --output-dir taxa -->
```

# 3.3 (Optional) Assess subset of taxonomic assignments with BLAST
```{bash}
qiime tools export \
    --input-path taxa/classification.qza \
    --output-path taxa
```


```{bash}
qiime feature-table tabulate-seqs \
    --i-data deblur_output/representative_sequences.qza \
    --o-visualization deblur_output/representative_sequences.qzv
```

# 4. Filtering resultant table
## 4.1. Filter out rare ASVs
```{bash}
qiime feature-table filter-features \
    --i-table deblur_output/table.qza \
    --p-min-frequency 11 \
    --p-min-samples 1 \
    --o-filtered-table deblur_output/deblur_table_filt.qza
```

## 4.2. Filter out contaminant and unclassified ASVs

```{bash}
qiime taxa filter-table \
    --i-table deblur_output/deblur_table_filt.qza \
    --i-taxonomy taxa/classification.qza \
    --p-include p__ --p-exclude mitochondria,chloroplast \
    --o-filtered-table deblur_output/deblur_table_filt_contam.qza
```

## 4.3. (Optional) Exclude low-depth samples
## 4.4. Subset and summarize filtered table
```{bash}
qiime feature-table summarize \
    --i-table deblur_output/deblur_table_filt_contam.qza \
    --o-visualization deblur_output/deblur_table_filt_contam_summary.qzv
```
Question 8: What is the maximum sequencing depth across all samples?
11690
```{bash}
cp deblur_output/deblur_table_filt_contam.qza deblur_output/deblur_table_final.qza
qiime feature-table filter-seqs \
    --i-data deblur_output/representative_sequences.qza \
    --i-table deblur_output/deblur_table_final.qza  \
    --o-filtered-data deblur_output/rep_seqs_final.qza
```

Finally, you can make a new summary of the final filtered abundance table:
```{bash}
qiime feature-table summarize \
    --i-table deblur_output/deblur_table_final.qza \
    --o-visualization deblur_output/deblur_table_final_summary.qzv
```


```{bash}
#first export the taxonomic classifications as a table
qiime tools export \
  --input-path taxa/classification.qza \
  --output-path taxa
```


```{bash}
#rename some of this table
sed -i -e '1 s/Feature/#Feature/' -e '1 s/Taxon/taxonomy/' taxa/taxonomy.tsv
```


```{bash}
#export the final table (by default this will be in .biom format) - we call this a feature table, and within it, 
#rows will be ASVs and columns will be samples, with each cell containing the number of reads assigned to an ASV within that sample
qiime tools export \
  --input-path deblur_output/deblur_table_final.qza \
  --output-path exports
```


```{bash}
#add the taxonomy information to the feature table
biom add-metadata \
  -i exports/feature-table.biom \
  -o exports/feature-table_w_tax.biom \
  --observation-metadata-fp taxa/taxonomy.tsv \
  --sc-separated taxonomy
```


```{bash}
#convert the .biom feature table with taxonomy into a .txt file
biom convert \
  -i exports/feature-table_w_tax.biom \
  -o exports/feature-table_w_tax.txt \
  --to-tsv \
  --header-key taxonomy
```


```{bash}
#export the ASV sequences as a fasta file
qiime tools export \
  --input-path deblur_output/rep_seqs_final.qza \
  --output-path exports
```


```{bash}
```


```{bash}
```


```{bash}
```


```{bash}
```


```{bash}
```


```{bash}
```


```{bash}
```

